xtra-libp2p/src/dialer.rs

```diff
@@ -96,18 +84,42 @@ impl Actor {
         self.listener_peer_id
             .expect("to always have peer id if successfully started")
     }
+
+    async fn is_connection_established(&self) -> Result<bool> {
+        Ok(self
+            .endpoint
+            .send(GetConnectionStats)
+            .await?
+            .connected_peers
+            .contains(&self.peer_id()))
+    }
+
+    async fn dial(&self) -> Result<()> {
+        anyhow::ensure!(
+            !self.is_connection_established().await?,
+            "Connection should not be active when dialing"
+        );
+
+        if let Err(e) = self.connect().await {
+            tracing::warn!("Failed to request connection from endpoint: {e:#}");
+        }
+
+        // Only check the connection again after it had enough time to be established
+        tokio::time::sleep(CONNECTION_TIMEOUT).await;
+
+        anyhow::ensure!(
+            self.is_connection_established().await?,
+            "No connection after dialing attempt",
+        );
+        Ok(())
+    }
 }
 
 #[xtra_productivity]
 impl Actor {
-    async fn handle(&mut self, _msg: StopIfNotConnected, ctx: &mut xtra::Context<Self>) {
-        if !self.connected {
-            self.stop_with_error(
-                Error::Failed {
-                    source: anyhow!("Did not connect in time"),
-                },
-                ctx,
-            )
+    async fn handle(&mut self, _msg: Dial, ctx: &mut xtra::Context<Self>) {
+        if let Err(e) = self.dial().await {
+            self.stop_with_error(Error::Failed { source: e }, ctx);
```

Comment:
Maybe add a log with the error here so we know what failed :)
Or let the supervisor log it.

Reply:
added inside `stop_with_error` function, both on dialer and listener.